src/catalog/catalog.cpp

```diff
@@ -30,421 +30,714 @@ Catalog *Catalog::GetInstance(void) {
+  InitializeCatalog();
 
   // Create metrics table in default database
+  // TODO: stats?
   CreateMetricsCatalog();
 
   InitializeFunctions();
 }
 
-void Catalog::CreateMetricsCatalog() {
-  auto default_db = GetDatabaseWithName(CATALOG_DATABASE_NAME);
-  auto default_db_oid = default_db->GetOid();
+void Catalog::InitializeCatalog() {
+  // Create pg_catalog database
+  auto pg_catalog = new storage::Database(CATALOG_DATABASE_OID);
+  pg_catalog->setDBName(CATALOG_DATABASE_NAME);
+  databases_.push_back(pg_catalog);
+
+  // Create catalog tables
+  auto pg_database = DatabaseCatalog::GetInstance(pg_catalog, pool_.get());
+  auto pg_table = TableCatalog::GetInstance(pg_catalog, pool_.get());
+  IndexCatalog::GetInstance(pg_catalog, pool_.get());
+  //  ColumnCatalog::GetInstance(); // Called implicitly
+
+  // Create indexes on catalog tables, insert them into pg_index
+  // note that CreateIndex() from catalog.cpp will create index on storage level
+  // table and at the same time insert a new index record into pg_index
+  // TODO: This should be hash index rather than tree index?? (but postgres use
+  // btree!!)
+
+  CreatePrimaryIndex(CATALOG_DATABASE_OID, DATABASE_CATALOG_OID);
+  CreatePrimaryIndex(CATALOG_DATABASE_OID, TABLE_CATALOG_OID);
+
+  CreateIndex(CATALOG_DATABASE_OID, DATABASE_CATALOG_OID,
+              std::vector<std::string>({"database_name"}),
+              DATABASE_CATALOG_NAME "_skey0", IndexType::BWTREE,
+              IndexConstraintType::UNIQUE, true, nullptr, true);
+
+  CreateIndex(CATALOG_DATABASE_OID, TABLE_CATALOG_OID,
+              std::vector<std::string>({"table_name", "database_oid"}),
+              TABLE_CATALOG_NAME "_skey0", IndexType::BWTREE,
+              IndexConstraintType::UNIQUE, true, nullptr, true);
+  CreateIndex(CATALOG_DATABASE_OID, TABLE_CATALOG_OID,
+              std::vector<std::string>({"database_oid"}),
+              TABLE_CATALOG_NAME "_skey1", IndexType::BWTREE,
+              IndexConstraintType::DEFAULT, false, nullptr, true);
+
+  // actual index already added in column_catalog, index_catalog constructor
+  // the reason we treat those two catalog tables differently is that indexes
+  // needs to be built before insert tuples into table
+  IndexCatalog::GetInstance()->InsertIndex(
+      COLUMN_CATALOG_PKEY_OID, COLUMN_CATALOG_NAME "_pkey", COLUMN_CATALOG_OID,
+      IndexType::BWTREE, IndexConstraintType::PRIMARY_KEY, true,
+      std::vector<oid_t>({0, 1}), pool_.get(), nullptr);
+  IndexCatalog::GetInstance()->InsertIndex(
+      COLUMN_CATALOG_SKEY0_OID, COLUMN_CATALOG_NAME "_skey0",
+      COLUMN_CATALOG_OID, IndexType::BWTREE, IndexConstraintType::UNIQUE, true,
+      std::vector<oid_t>({0, 2}), pool_.get(), nullptr);
+  IndexCatalog::GetInstance()->InsertIndex(
+      COLUMN_CATALOG_SKEY1_OID, COLUMN_CATALOG_NAME "_skey1",
+      COLUMN_CATALOG_OID, IndexType::BWTREE, IndexConstraintType::DEFAULT,
+      false, std::vector<oid_t>({0}), pool_.get(), nullptr);
```

Comment:
Should the updates in IndexCatalog and the CreateIndex be placed in the same transaction? Would it be a problem if CreateIndex succeeds but InsertIndex fails because of transaction aborts?